src/main/java/org/semux/core/Wallet.java

```diff
@@ -102,13 +117,22 @@ public boolean unlock(String password) {
 
         try {
             byte[] key;
+            byte[] salt;
 
             if (exists()) {
                 SimpleDecoder dec = new SimpleDecoder(IOUtil.readFile(file));
                 int version = dec.readInt(); // version
 
-                List<Key> newAccounts;
+                Set<Key> newAccounts;
                 Map<ByteArray, String> newAliases = new HashMap<>();
+
+                //initialize new HD wallet if not created.
+                if(version < 4) {
+                    logger.info("Creating new HD Wallet!");
+                    hdSeed = new byte[64];
```

Comment:
What is a `hdSeed` when version < 4? What is a purpose of creating a random seed when opening an old wallet?

Reply:
So, when I open a v3 wallet.data in Semux, it will generate new seed for further new accounts, and will upgrade the wallet file to v4 with a new seed and old non-seed generated accounts?

Reply:
Correct.  And if you lose your wallet, you can take that seed and restore all the addresses you created after upgrade (but not any pre-existing addresses).  We'll have to work on user messaging so this is clear